lib/measured/conversion_table_builder.rb

```diff
@@ -26,20 +26,34 @@ def to_h
   private
 
   def find_conversion(to:, from:)
-    conversion = find_direct_conversion(to: to, from: from) || find_tree_traversal_conversion(to: to, from: from)
+    conversion = find_direct_conversion_cached(to: to, from: from) || find_tree_traversal_conversion(to: to, from: from)
 
     raise Measured::UnitError, "Cannot find conversion path from #{ from } to #{ to }." unless conversion
 
     conversion
   end
 
+  def find_direct_conversion_cached(to:, from:)
+    @cache ||= {}
+    @cache[to] ||= {}
+    cached_value = @cache[to][from]
+
+    if cached_value == nil
+      @cache[to][from] = find_direct_conversion(to: to, from: from)
+    elsif cached_value == false
+      nil
+    else
+      cached_value
+    end
+  end
```

Comment:
If only we had a `Maybe` type in Ruby...

I'll propose another idea here, because this code is a little challenging to decipher when you first see it.

Consider that `find_conversion` is basically the main branching behaviour in `to_h`: either we get a direct conversion or try to find a path that gives us one. Also, more importantly, note that this branching behaviour doesn't change through the loop.

Given that, why not build the table in two passes:

1. compute all of the direct conversions, then
2. for everything without a conversion value, compute the tree traversal conversions.

Reply:
> I'll propose another idea here because this code is a little challenging to decipher when you first see it.

I totally agree, the code right now is not very readable

>Given that, why not build the table in two passes:
>
> 1. compute all of the direct conversions, then
> 2. for everything without a conversion value, compute the tree traversal conversions.

I like the idea, but I'm not sure what would be best 😕 . If I understood you correctly we would generate all the conversions before, like in the constructor method, right? 

On one hand, some approach like this will potentially cache data that will not be used even once. It will also take some time to generate that cache even if we don't use it all. On the other hand, it will provide more consistent timings, as we wouldn't need to compute the cache on the first call

Reply:
Not quite, and no worries about my proposed idea because this is already a great improvement 🎉